### ISSUE TITLE
Buffer overflow in Coords_adjacent

### ISSUE THREAD

**Issue by edre** [richardjs/zoe#15]
```
$ CC='clang -fsanitize=address' make all
$ ./bench
MCTS, 1 iteration...=================================================================
==7695==ERROR: AddressSanitizer: global-buffer-overflow on address 0x5582d59edc78 at pc 0x55
82d4fe8053 bp 0x7ffde57a2db0 sp 0x7ffde57a2da8
READ of size 1 at 0x5582d59edc78 thread T0
    #0 0x5582d4fe8052 in Coords_adjacent (/home/edre/code/games/zoe/src/bench+0xda052) (BuildId: c89802feb071681b712ee259a07d15b2622e63c1)
    #1 0x5582d4ff080b in State_add_action (/home/edre/code/games/zoe/src/bench+0xe280b) (BuildId: c89802feb071681b712ee259a07d15b2622e63c1)
    #2 0x5582d4ffc636 in State_derive_actions (/home/edre/code/games/zoe/src/bench+0xee636 (BuildId: c89802feb071681b712ee259a07d15b2622e63c1)
    #3 0x5582d4fff944 in State_act (/home/edre/code/games/zoe/src/bench+0xf1944) (BuildId: c89802feb071681b712ee259a07d15b2622e63c1)
    #4 0x5582d4fee5ec in State_simulate (/home/edre/code/games/zoe/src/bench+0xe05ec) (BuildId: c89802feb071681b712ee259a07d15b2622e63c1)
    #5 0x5582d4fe9a18 in iterate (/home/edre/code/games/zoe/src/bench+0xdba18) (BuildId: c89802feb071681b712ee259a07d15b2622e63c1)
    #6 0x5582d4fe9f70 in mcts (/home/edre/code/games/zoe/src/bench+0xdbf70) (BuildId: c89802feb071681b712ee259a07d15b2622e63c1)
    #7 0x5582d4f2b5ad in main (/home/edre/code/games/zoe/src/bench+0x1d5ad) (BuildId: c89802feb071681b712ee259a07d15b2622e63c1)
    #8 0x7f8b7ba39aac  (/lib/ld-musl-x86_64.so.1+0x1baac)

0x5582d59edc78 is located 24 bytes after global variable 'coords_adjacent_map' defined in 'coords.c' (0x5582d599cc60) of size 331776
```

I see workers segfault when I try to play a full game, this might be related. The main program stalls waiting for input which never comes.

**Comment by richardjs** [comment]
Turns out I wasn't initializing the map of adjacent coords in the UHP code.

The whole coords system is something I hope to review and clean up, but in the meantime adding a call to `init_coords`in `zoe_uhp.c` seems to have fixed it. I'm also curious this also fixes #16, because this is one of those "how was it even working at all?" moments.

**Comment by edre** [comment]
The original report was from the `bench` command, and I can still reproduce that with this change. I don't think it uses UHP code.

With this change I can now consistently reproduce a stacktrace closer to this one from the `zoe_uhp` binary.

```
$ CC='clang -fsanitize=address' make zoe_uhp
$ nokamute --num-threads=1 play  --game-type=Base ./zoe_uhp ai
...
AddressSanitizer: global-buffer-overflow on address 0x55b811b6bc78 at pc 0x55b811165d13 bp 0x7ffdec93b550 sp 0x7ffdec93b548
READ of size 1 at 0x55b811b6bc78 thread T0
    #0 0x55b811165d12 in Coords_adjacent (/home/edre/code/games/zoe/src/zoe_uhp+0xd9d12) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #1 0x55b81116e4cb in State_add_action (/home/edre/code/games/zoe/src/zoe_uhp+0xe24cb) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #2 0x55b81117a2f6 in State_derive_actions (/home/edre/code/games/zoe/src/zoe_uhp+0xee2f6) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #3 0x55b81117d604 in State_act (/home/edre/code/games/zoe/src/zoe_uhp+0xf1604) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #4 0x55b81116c2ac in State_simulate (/home/edre/code/games/zoe/src/zoe_uhp+0xe02ac) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #5 0x55b8111676d8 in iterate (/home/edre/code/games/zoe/src/zoe_uhp+0xdb6d8) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #6 0x55b81116777f in iterate (/home/edre/code/games/zoe/src/zoe_uhp+0xdb77f) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #7 0x55b81116777f in iterate (/home/edre/code/games/zoe/src/zoe_uhp+0xdb77f) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #8 0x55b811167c30 in mcts (/home/edre/code/games/zoe/src/zoe_uhp+0xdbc30) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #9 0x55b811186be1 in think (/home/edre/code/games/zoe/src/zoe_uhp+0xfabe1) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #10 0x55b81118b1e8 in bestmove (/home/edre/code/games/zoe/src/zoe_uhp+0xff1e8) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #11 0x55b81118bac7 in uhp_loop (/home/edre/code/games/zoe/src/zoe_uhp+0xffac7) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #12 0x55b811164c0e in main (/home/edre/code/games/zoe/src/zoe_uhp+0xd8c0e) (BuildId: 8f0a8fb43fbe50980a2c923190c6348d1cdb79c7)
    #13 0x7fcfed9c6aac  (/lib/ld-musl-x86_64.so.1+0x1baac)

0x55b811b6bc78 is located 24 bytes after global variable 'coords_adjacent_map' defined in 'coords.c' (0x55b811b1ac60) of size 331776
```

**Comment by richardjs** [comment]
Sorry, my mistake. I thought I had ran your test with `./bench` and gotten no errors, but I neglected to force make to recompile everything with the new `CC`. I can confirm your original issue is still in place.

**Comment by richardjs** [comment]
OK, I think I got it this time.

Point 1: [`Coords_adjacent` calls `coords_map_adjacent`](https://github.com/richardjs/zoe/blob/c47e271e226e37ffa860319a2bca9ba4131ba3a2/src/coords.c#L101-L109), which returns the results from a precomputed adjacency matrix. The matrix is a 4D array, with dimensions for the Q and R of each of the two coordinates. (The precomputed matrix was an experiment--benchmarks suggested it gave a small speed improvement, but I never felt quite right about it.)

Point 2: I represent a move (`struct Action`) as a pair of coordinates: the source coordinates, and the destination coordinates. Place actions are a special case: the source Q is set to the special value `PLACE_ACTION`, and the source R is set to the piece type. This keeps actions compact, but it occasionally causes problems when code assumes that given coordinates refer to a place on the hex grid.

Given those two points, it's easy to guess the issue. Evidently there are places I'm passing a source coordinate from a place action to `Coords_adjacent`. The Q value isn't a proper index for the adjacency matrix, and so it breaks.

Rather than making sure I never call `Coords_adjacent` with place coords, I fixed `Coords_adjacent` to properly handle place action coords.

I've tested it with your commands (properly this time... I hope) and they seem to work. Let me know what you see. And thanks for your help with all this!
